multicluster/controllers/multicluster/stale_controller.go

```diff
@@ -102,128 +122,169 @@ func (c *StaleController) cleanupStaleServiceResources(remoteCluster commonarea.
-		},
+func (c *StaleResCleanupController) cleanupACNPResources(resImpList *mcsv1alpha1.ResourceImportList) error {
+	acnpList := &crdv1alpha1.ClusterNetworkPolicyList{}
+	if err := c.List(ctx, acnpList, &client.ListOptions{}); err != nil {
+		return err
+	}
+	staleMCACNPItems := map[string]crdv1alpha1.ClusterNetworkPolicy{}
+	for _, acnp := range acnpList.Items {
+		if _, ok := acnp.Annotations[common.AntreaMCACNPAnnotation]; ok {
+			staleMCACNPItems[acnp.Name] = acnp
+		}
+	}
+	for _, resImp := range resImpList.Items {
+		if resImp.Spec.Kind == common.AntreaClusterNetworkPolicyKind {
+			acnpNameFromResImp := common.AntreaMCSPrefix + resImp.Spec.Name
+			delete(staleMCACNPItems, acnpNameFromResImp)
+		}
+	}
+	for _, stalePolicy := range staleMCACNPItems {
+		acnp := stalePolicy
+		klog.InfoS("Cleaning up stale ACNP", "acnp", klog.KObj(&acnp))
+		if err := c.Client.Delete(ctx, &acnp, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
 	}
-	selector, _ := metav1.LabelSelectorAsSelector(&labelSelector)
-	if err := remoteCluster.List(ctx, resExpList, &client.ListOptions{Namespace: remoteCluster.GetNamespace(), LabelSelector: selector}); err != nil {
+	return nil
+}
+
+func (c *StaleResCleanupController) cleanupClusterInfoImport(resImpList *mcsv1alpha1.ResourceImportList) error {
+	ciImpList := &mcsv1alpha1.ClusterInfoImportList{}
+	if err := c.List(ctx, ciImpList, &client.ListOptions{}); err != nil {
 		return err
 	}
+
+	staleCIImps := map[string]mcsv1alpha1.ClusterInfoImport{}
+	for _, item := range ciImpList.Items {
+		staleCIImps[item.Name] = item
+	}
+	for _, resImp := range resImpList.Items {
+		if resImp.Spec.Kind == common.ClusterInfoKind {
+			delete(staleCIImps, resImp.Name)
+		}
+	}
+	for _, staleCIImp := range staleCIImps {
+		ciImp := staleCIImp
+		klog.InfoS("Cleaning up stale ClusterInfoImport", "clusterinfoimport", klog.KObj(&ciImp))
+		if err := c.Client.Delete(ctx, &ciImp, &client.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+// cleanupServiceResourceExport removes any Service/Endpoint kind of ResourceExports when there is no
+// corresponding ServiceExport in the local cluster.
+func (c *StaleResCleanupController) cleanupServiceResourceExport(commonArea commonarea.RemoteCommonArea,
+	resExpList *mcsv1alpha1.ResourceExportList) error {
 	svcExpList := &k8smcsv1alpha1.ServiceExportList{}
 	if err := c.List(ctx, svcExpList, &client.ListOptions{}); err != nil {
 		return err
 	}
-	resExpItems := resExpList.Items
+	allResExpItems := resExpList.Items
 	svcExpItems := svcExpList.Items
+	var staleResExpItems []mcsv1alpha1.ResourceExport
 
-	for k, re := range resExpItems {
+	for i, resExp := range allResExpItems {
+		if (resExp.Spec.Kind == common.ServiceKind || resExp.Spec.Kind == common.EndpointsKind) && resExp.Labels[common.SourceClusterID] == c.localClusterID {
+			staleResExpItems = append(staleResExpItems, allResExpItems[i])
+		}
+	}
```

Comment:
The following loop can be reduced to O(n) by having two maps: staleServiceExports, staleEndpointsExports

Reply:
done